dbms/src/Interpreters/ExternalLoader.cpp

```diff
@@ -747,6 +821,12 @@ class ExternalLoader::LoadingDispatcher : private boost::noncopyable
         if (!new_object && !new_exception)
             throw Exception("No object created and no exception raised for " + type_name, ErrorCodes::LOGICAL_ERROR);
 
+        /// Lock the mutex again to store the changes.
+        if (async)
+            lock.lock();
```

Comment:
Dangerous logic

Reply:
Agree with you, but I don't know how to make it safer. I just moved this `if` upper, because there was a race condition in `calculateNextUpdateTime`.

Reply:
Refactored in https://github.com/ClickHouse/ClickHouse/pull/7360/commits/ae42dd1cea1d130299a8d54546dab4611a5c3d92